[Bug Fix] First usage of terminalwriter creates an unexpected new line. closes #12777

Oreldm opened the pull request:
- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [x] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [X] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/main/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [X] Add yourself to `AUTHORS` in alphabetical order.

Oreldm commented:
I can't edit labels to add type:bug . 
This is the issue: https://github.com/pytest-dev/pytest/issues/12777

dongfangtianyu commented:
Thank you  @Oreldm for PR.
Could you please add a test case for this? 
It will not only clearly demonstrate the effectiveness of this PR but also help prevent regressions caused by other PRs.

Oreldm commented:
Thanks @dongfangtianyu . I added a test case for this.
Wasn't trivial as I needed to reproduce the problem during the test but I finally managed to do it.

nicoddemus commented on changelog/12777.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Fixed bug where the first writeline in terminalwriter created a new line.
```
```suggestion
Fixed bug where the `TerminalWriter` created an extraneous newline line in some situations.
```

nicoddemus commented on testing/test_terminal.py:
```diff
@@ -130,6 +130,52 @@ def test_writeline(self, pytester: Pytester, linecomp) -> None:
+        self, pytester: Pytester, linecomp
+    ) -> None:
+        r"""
+        This test is checking issue 12777 - where writeline is writing a new line at the first usage, due to
```
Thanks @Oreldm for taking the time to write this, but I feel this is testing too much of the internals, which makes this test brittle and hard to maintain.

Can you try to reproduce the problem just via a `pytester.runpytest` call and checking the output? It is of course important that this test without your change fails, and passes with it.

Oreldm replied:
Thanks @nicoddemus . I agree, it felt wrong for me as well. I refactored the test.

Oreldm commented:
I'm testing %100 of my changes, why is codecov thinks its only 95.23?
It might not concider code that runs from a string- such as: pytester.makeconftest(rf"""
            import pytest
 ... " as coverage?

Oreldm commented:
Hi @nicoddemus . 
During my testing I found another case that the bug will happen so I changed my code accordingly. 

I can see that codecov/patch is on %95.00 although I'm testing all the code I added.
Do you know why? Should I fix it? Not sure how (I guess its because codecov/patch is not considering the code string inside pytester as coverage?). 

Thanks!

nicoddemus commented on src/_pytest/terminal.py:
```diff
@@ -588,7 +596,8 @@ def pytest_runtest_logstart(
     def pytest_runtest_logreport(self, report: TestReport) -> None:
-        self._tests_ran = True
+        if report.when == "teardown":
+            self._all_tests_finished = True
```
Note that `pytest_runtest_logreport` is not a good heuristic to set `_all_tests_finished` to True:

* This hook is called multiple times for each test
* Even for the teardown event, this might be called multiple times per test (for example pytest-repeat does this).

TBH I'm not a big fan on how this is turning out, IME every time state is being represented by a bunch of "control flags" like in here means a maintenance problem later... but I do not have a concrete suggestion either, the fact that you had to use those control flags suggest that the entire code needs a proper rethinking, giving it just grown organically over the years...

What do you think?

Oreldm replied:
I will be honest- I agree. 
After hours of thinking this was the simplest solution I found. 
So.. it fixes the bug completly- but at what cost? Also- the current bug started from the exact reason you just mentioned- we are update currentfspath control flag, which is kind of an anti pattern- or trying to escape from design. In the current implementation we are sometimes updating currentfsflag to -2 (what does it mean? What is this magical number?).

I can go to the journey of entire refactor of the code to fix it, but Im new here, so I tried to keep things simple. 
What do you think?

nicoddemus replied:
@Oreldm I'm relieved we are on the same page. :sweat_smile: 

Not only this is critical section of the code (so any bugs here will reach almost any pytest run out there), I know there are plugins and IDEs that access those internals today (not to their fault, we do not have a public API and sometimes this is the best they can do), so any refactoring here is really risky.

TBH, I would abort this and leave the bug in place for now, it is nothing critical and the solution to refactor everything is not worth the risk I think.

nicoddemus replied:
Sorry you have wasted so much time here, but in the end I think it is better to leave this bug alone for now.

Oreldm replied:
Well... Unlucky start for me, but maybe next time :) Aborted

nicoddemus replied:
Yeah please do not get discouraged to contribute... definitely bad luck. 😁 

Cheers and thanks for being so understanding.
